### DISCUSSION TITLE
How to configure Wagmi when using Ethereum Gas Station Network for gasless txns?

### DISCUSSION THREAD

**Discussion by feliche93** [wevm/wagmi#958] [Q&A]
Hey frens,

**TLDR:** I'm working with @h4gen on integrating GSN into our smart contract. We are currently struggling to understand how we can integrate the GSN Provider into wagmi and call the `usePrepareContractWrite` hook.

GSN provides the following documentation on setting it up in JS: https://docs-v2.opengsn.org/javascript-client/getting-started.html#add-use-gsn-relayprovider-in-your-app

```
const { RelayProvider } = require('@opengsn/provider')

const config = { 
    paymasterAddress,
    loggerConfiguration: {
        logLevel: 'debug',
        // loggerUrl: 'logger.opengsn.org',
    }
}
const provider = await RelayProvider.newProvider({ provider: web3.currentProvider, config }).init()
const web3 = new Web3(provider);
```

Later on this is how you use it for txns:

```
const myRecipient = new web3.eth.Contract(abi, address);

// Sends the transaction via the GSN
await myRecipient.methods.myFunction().send({ from });

// Disable GSN for a specific transaction (but require that the sender has eth!)
await myRecipient.methods.myFunction().send({ from, useGSN: false });
```

We don't know how to get this as a native provvider into WAGMI and how to call it with the right hooks. Any examples, hints to docu or tutorials would be very appreciated!

Thanks a lot in advance!

**Comment by feliche93** [comment]
Little update, in the meantime we found that there's a wrapper you can use:

```
const gsnContract = () => wrapContract(contract.contracts.ChainTagV0, gsnConfig);
const gsnSigner = () => wrapSigner(signer, gsnConfig);
```

How would we use this with the existing WAGMI hooks?

**Reply by c-neri** [reply]
I created a simple connector here: https://github.com/c-neri/GSN/blob/master/app/connectors/GSN.ts. So it return a wrapped signer.How ever i got a strange error when i submit the txs.
